### PR TITLE
Fixing symcc custom mutator

### PR DESCRIPTION
**Pull request by Rumata888**
Changed symcc custom mutator to work both with stdin and file-input targets and fixed several errors in the mutator. Added two example targets for easy tests in the future. Added printing fo enqueued items (otherwise afl seems stuck during initial processing of huge input corpus). Fixed the problem where symcc overwrote previous testcases.

### CODE REVIEW

File: custom_mutators/symcc/symcc.c [hunk 1]
```diff
@@ -95,40 +98,66 @@ void afl_custom_queue_new_entry(my_mutator_t * data,
                                 const uint8_t *filename_new_queue,
                                 const uint8_t *filename_orig_queue) {
 
+  int pipefd[2];
+  struct stat st;
+  ACTF("Queueing to symcc: %s", filename_new_queue);
+  u8 *fn = alloc_printf("%s", filename_new_queue);
+  if (!(stat(fn, &st) == 0 && S_ISREG(st.st_mode) && st.st_size)) {
+    PFATAL("Couldn't find enqueued file: %s",fn);
```
Comment (hexcoder-) [inline custom_mutators/symcc/symcc.c]: mem leak for `fn`

File: custom_mutators/symcc/symcc.c [hunk 2]
```diff
@@ -95,40 +98,66 @@ void afl_custom_queue_new_entry(my_mutator_t * data,
...
+  }
+   
+  if (afl_struct->fsrv.use_stdin){
+    if (pipe(pipefd)==-1)
+    {
+      exit(-1);
```
Comment (hexcoder-) [inline custom_mutators/symcc/symcc.c]: better use PFATAL and tell why we have to abort.

File: custom_mutators/symcc/symcc.c [hunk 3]
```diff
@@ -95,40 +98,66 @@ void afl_custom_queue_new_entry(my_mutator_t * data,
...
         if (r <= 0) return;
-        close(0);
-        ck_write(0, data->mutator_buf, r, fn);
-        ck_free(fn);
-
+        close(fd);
+        if (r>fcntl(pipefd[1],F_GETPIPE_SZ)) fcntl(pipefd[1],F_SETPIPE_SZ,MAX_FILE);
+        ck_write(pipefd[1], data->mutator_buf, r, filename_new_queue);
+      } else {
+        PFATAL("Something happened to the enqueued file before sending its contents to symcc binary");
```
Comment (hexcoder-) [inline custom_mutators/symcc/symcc.c]: mem leak for `fn`

**Comment by hexcoder-** [comment]
Thanks for your PR!

**Comment by Rumata888** [comment]
Ok, I'll fix the memleaks and resubmit it. Thanks for auditing it.

**Comment by Rumata888** [comment]
Added the fixes.

**Comment by vanhauser-thc** [comment]
IMHO this looks fine. @hexcoder- ?

File: custom_mutators/symcc/symcc.c [hunk 4]
```diff
@@ -135,6 +141,7 @@ void afl_custom_queue_new_entry(my_mutator_t * data,
 
         ssize_t r = read(fd, data->mutator_buf, MAX_FILE);
         DBG("fn=%s, fd=%d, size=%ld\n", fn, fd, r);
+        ck_free(fn);
         if (r <= 0) return;
         close(fd);
```
Comment (hexcoder-) [inline custom_mutators/symcc/symcc.c:145]: better move close(fd) one line up before the return to avoid leaking fd.

Comment (vanhauser-thc) [reply custom_mutators/symcc/symcc.c:145]: good catch!

Comment (hexcoder-) [reply custom_mutators/symcc/symcc.c:145]: If we can get this fix as well, I am happy.

**Comment by Rumata888** [comment]
Ok. Do you notice anything else or is that it?

**Comment by Rumata888** [comment]
Updated it (also closed pipe before return and PFATAL)

**Comment by vanhauser-thc** [comment]
thanks!
the pipe close before the fatal is unnecessary as fatal exits and then closes all file descriptors anyway, but it doesnt hurt :)

**Pull request merged**